GC Thread Pool Tuning for CRIU

RSalman opened the pull request:
These changes aim to optimize/tune GC threading footprint observed when using CRIU (Checkpoint/Restore) capabilities. Tests have shown excessive number of GC threads to have a significant impact on checkpoint image footprint and restore times. Contracting the GC thread pool during the checkpoint phase and expanding it during restore phase has shown to be a better strategy.

These changes introduce APIs to adjust the size of the GC (dispatcher) thread pool and make adjustments to dependent parties consuming the thread count:
- Introduced `contractThreadPool` (to shutdownThreads for checkpoint) and expandThreadPool (to startupThreads for restore)
- Introduced `checkpointGCthreadCount` in extensions (user adjustable via cmd line opts), consumed to contract thread pool.
- Refactored and introduced `supportedGCThreadCount` API to return number of GC threads supported based on hardware and dispatcher's max
- `_threadCountMaximumAtCheckpointStartup` introduced in dispatcher to take snapshot of initial thread pool size, this is used as hard cap on # of threads during restore.
- Exisiting `startupThreads()` main routine refactored into `internalStartupThreads` for reusability with expandThreadPool.
- New MM Trace Point introduced to log `contractThreadPool` and `expandThreadPool`, all trc points configured as `noEnv Overhead=1 Level=1 Group=dispatcher`
- Introduced  `adjustGCThreadCountOnCheckpoint` & `reinitializeGCThreadCountOnRestore` in `MM_Configuration` as general high level entry point to coordinate  checkpoint/restore between GC subcomponents
  - Overide `reinitializeGCThreadCountOnRestore` for `MM_ConfigurationGenerational` to make Scavenger specific adjustments on thread count change.

Signed-off-by: Salman Rana <salman.rana@ibm.com>

RSalman commented:
@dmitripivkine, @amicic and I had a discussion on how the thread sizing should be done in this iteration, I've outlined it below, @tajila @vijaysun-omr please have a look. 

**The thread pool is contracted and expanded as follows:**
- Initial startup thread pool sizing remains unchanged. i.e Thread count is either user specified (Xgcthreads) or default # CPUs.
- Checkpoint thread count defaults to 4,  (can be specified by the user -XXgc:checkpointGCThreadcount=…)
  - Checkpoint thread count must be less than initial startup thread count
  - If the initial thread pool size is < 4, then no action is taken in default case. However, user specified count will always be obeyed (e.g user sets checkpoint thread count to 1)
- Restore thread count defaults to # CPU on restore machine (can be specified by the user -XXgc:restoreGCThreadcount=…)
  - Restore thread count is capped by initial startup thread pool size (can’t exceed the initial thread count from time of VM startup)
  - If user specified restore count > Initial VM startup thread count, then fall back to # CPU on restore machine
  - If # CPU on restore machine > Initial VM startup thread count, then no action is taken

**Examples/Test cases which illustrate these rules:**
<img width="853" alt="image" src="https://user-images.githubusercontent.com/9648788/204850083-5d063368-e861-43ed-9568-9f5fc386355d.png">

**Performance Data:**
![image](https://user-images.githubusercontent.com/9648788/204850745-da480d12-f076-469f-8662-37e9306ec72a.png)

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+
+		uintptr_t expectedThreadShutdownThread = newThreadCount - 1;
+
+		omrthread_monitor_enter(_dispatcherMonitor);
```
not sure, but suspect you don't really need to exit and re-enter monitor

amicic replied:
my bad, it's not the same monitor

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+MM_ParallelDispatcher::contractThreadPool(MM_EnvironmentBase *env, uintptr_t newThreadCount)
+{
+	Assert_MM_false(_workerThreadsReservedForGC);
+	Assert_MM_false(_inShutdown);
```
do we end up calling shutDownThreads after this?
either way, we should probably assert the same at the start of shutDownThreads

RSalman replied:
shutDownThreads should only be called during VM shutdown after this point

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = _threadCount = _threadCountMaximum = newThreadCount;
+		_activeThreadCount = adjustThreadCount(_threadCount);
```
it's probably simpler to just set this to _threadCountMaximum, too.
it will be adjusted to a meaningful value at restore time, when it's really to be used, anyway

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+
+		Assert_MM_true(_threadShutdownCount == expectedThreadShutdownThread);
+
+		_extensions->gcThreadCount = _threadCount = _threadCountMaximum = newThreadCount;
```
break this into separate lines

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -258,7 +259,8 @@ MM_ParallelDispatcher::initialize(MM_EnvironmentBase *env)
 	OMR::GC::Forge *forge = env->getForge();
 
-	_threadCountMaximum = env->getExtensions()->gcThreadCount;
+	_threadCountMaximum = _threadCountAtGCInit = env->getExtensions()->gcThreadCount;
```
break this into separate lines

I'd give _threadCountAtGCInit  a more specific name like
_threadCountMaximumAtSnapshotStartup
and surround it withe CRIU #ifdef 
(I can see that variable is used in a trace point - not sure if generation of tracepoints is aware of #ifdefs)

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+	 * 1) User didn't specify a restore thread count (restoreGCthreadCount == UDATA_MAX), or
+	 * 2) User specified a restore thread count, but it's greater than our hard limit (_threadCountAtGCInit) */
+	if (newThreadCount > _threadCountAtGCInit) {
+		newThreadCount = supportedGCThreadCount;
```
this line alone does not make sense (although compound effect with the next line is overall correct)
here we should simply cap to _threadCountAtGCInit
and then on the next line we should again cap to supportedGCThreadCount

but then if we get rid of restoreGCThreadCount, this method should not even pass newThread count and this should be:
set new thread count to _threadCountAtGCInit
cap to supportedGCThreadCount

RSalman commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +644,140 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		} else {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed(_threadShutdownCount);
+			/* We don't know how many threads started up prior to failing, infer from _threadShutdownCount */
+			newThreadCount = _threadShutdownCount + 1;
```
Should we return false and force failure when we fail to startup all threads, or just continue on (as this code does)?

Currently, during VM startup, if this occurs we shutdown all threads and return false to caller which handles it by returning error.

```
	bool result = internalStartupThreads(workerThreadCount, _threadCountMaximum);

	if (result) {
		_threadCount = _threadCountMaximum;
		_activeThreadCount = adjustThreadCount(_threadCount);
	} else {
		/* Clean up the thread table and monitors */
		shutDownThreads();
	}
```

RSalman replied:
In OpenJ9, it seems to shutdown collector and report an error 

```C++
int
gcStartupHeapManagement(J9JavaVM *javaVM)
...

	if (!extensions->dispatcher->startUpThreads()) {
		extensions->dispatcher->shutDownThreads();
		result = JNI_ENOMEM;
	}

	if (JNI_OK != result) {
		PORT_ACCESS_FROM_JAVAVM(javaVM);
		extensions->getGlobalCollector()->collectorShutdown(extensions);
		j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_FAILED_TO_STARTUP_GARBAGE_COLLECTOR);
		return result;
	}
```

amicic replied:
do the same, shutdown threads and return error to the caller (which probably should continue with failure path, too)

RSalman replied:
Will leave this code as is, i.e not assume we'll fail and leave it to the caller to take action based on the `false` value returned.

CRIU code will be changed: https://github.com/eclipse-openj9/openj9/pull/16385#issuecomment-1333142859

RSalman commented:
Following changes in latest commit to address review comments:
- Introduced  `adjustGCThreadCountOnCheckpoint` &
`reinitializeGCThreadCountOnRestore` in `MM_Configuration` as general high level entry point to coordinate  checkpoint/restore between GC subcomponents
  - Overide `reinitializeGCThreadCountOnRestore` for `MM_ConfigurationGenerational` to make Scavenger specific adjustments on thread count change.
- Removed `gcThreadPoolAdjusted` handlers and `J9HOOK_MM_PRIVATE_GC_THREAD_POOL_ADJUSTED` event and triggers. Not required as Configuration will call subcomponents on restore.

- Removed `restoreGCthreadCount`: not required now that the user option is removed in downstream OpenJ9

- Renamed `_threadCountAtGCInit` -> `_threadCountMaximumAtCheckpointStartup`

amicic commented on gc/base/standard/ConfigurationStandard.cpp:
```diff
@@ -98,14 +98,20 @@ MM_ConfigurationStandard::initialize(MM_EnvironmentBase* env)
 	MM_Configuration::initializeGCThreadCount(env);
 
 #if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	initializeConcurrentScavengerThreadCount(env);
```
an existing problem, but still: this code (now a method) should really exist in ConfigurationGenerational (hence no need to even be virtual)

RSalman replied:
`initializeGCThreadCount` and `initializeConcurrentScavengerThreadCount`
moved from `MM_ConfigurationStandard`  -> `MM_ConfigurationGenerational`

amicic commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +650,134 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Success();
+			Assert_MM_true(_threadShutdownCount == (newThreadCount - 1));
+		} else {
+			Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed(_threadShutdownCount);
```
you could move those 2 trace points (and actually have only one reporting result and whatever relevant internals you want) inside internalStartupThreads. it would be interesting in non-restore startup potential (albeit far less likely) failure, as well

I'd drop _criu_ from the name of trace points

RSalman replied:
Made the suggested changes along with few other things after revisiting trace points:

- Removed "criu" from trc point names/template descriptions
- Changed Trc point group identifier from "criudispatcher" to "dispatcher"
- Removed trace points:
`Trc_MM_ParallelDispatcher_criu_expandThreadPool_Failed` and
`Trc_MM_ParallelDispatcher_criu_expandThreadPool_Success`.
- Introduced new trace point `Trc_MM_ParallelDispatcher_internalStartupThreads_Failed` to log start up thread failure.
- Cleaned up template descriptions for consistency (Entry/Exit trace points description aligned for both checkpoint and restore trc points)
- Renamed `preRestoreThreadCount` variable to `preExpandThreadCount`

babsingh commented:
jenkins build all

Running PR builds for preliminary testing. @RSalman Can you squash the commits and remove the use of markdown from the commit message(s)?

RSalman commented:
All PR builds finished and passed (all checks passed, 18 successful checks)

amicic commented:
PR comment is a bit out of date (still refers to a new hook that we meanwhile removed)

babsingh commented on gc/base/standard/ConfigurationGenerational.hpp:
```diff
@@ -64,8 +74,18 @@ class MM_ConfigurationGenerational : public MM_ConfigurationStandard
 	bool initialize(MM_EnvironmentBase* env);
 	MM_MemorySubSpaceSemiSpace *createSemiSpace(MM_EnvironmentBase *envBase, MM_Heap *heap, MM_Scavenger *scavenger, MM_InitializationParameters *parameters, UDATA numaNode = UDATA_MAX);
 	virtual void tearDown(MM_EnvironmentBase* env);
+	/**
+	 * Sets the number of gc threads
+	 *
+	 * @param env[in] - the current environment
+	 */
+	virtual void initializeGCThreadCount(MM_EnvironmentBase* env);
```
Use the correct formatting for the function description: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly
```suggestion
	/**
	 * Sets the number of GC threads
	 *
	 * @param[in] env the current environment
	 * @return void
	 */
	virtual void initializeGCThreadCount(MM_EnvironmentBase* env);
```

babsingh commented on gc/base/standard/ConfigurationGenerational.hpp:
```diff
@@ -64,8 +74,18 @@ class MM_ConfigurationGenerational : public MM_ConfigurationStandard
 	uintptr_t calculateDefaultRegionSize(MM_EnvironmentBase *env);
+
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	void initializeConcurrentScavengerThreadCount(MM_EnvironmentBase* env);
```
Missing function description: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly

babsingh commented on gc/base/standard/Scavenger.hpp:
```diff
@@ -600,6 +600,8 @@ class MM_Scavenger : public MM_Collector
 	 */
 	void calculateRecommendedWorkingThreads(MM_EnvironmentStandard *env);
 
+	void resetRecommendedThreads() { _recommendedThreads = UDATA_MAX; };
```
missing function description: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly

babsingh commented on gc/base/standard/ConfigurationGenerational.hpp:
```diff
@@ -64,8 +74,18 @@ class MM_ConfigurationGenerational : public MM_ConfigurationStandard
+
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	void initializeConcurrentScavengerThreadCount(MM_EnvironmentBase* env);
+#endif
```
missing endif comment: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-conditional-compilation-directives-if-ifdef-ifndef-else-elif-endif
```suggestion
#endif /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
```

babsingh commented on gc/base/standard/ConfigurationGenerational.hpp:
```diff
@@ -54,6 +54,16 @@ class MM_ConfigurationGenerational : public MM_ConfigurationStandard
 	
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	/**
+	 * Startup gc threads on restore
+	 *
+	 * @param env[in] - the current environment
+	 */
```
Use the correct formatting for the function description: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#comment-function-parameters-clearly
```suggestion
	/**
	 * Startup GC threads on restore
	 *
	 * @param[in] env the current environment
	 * @return void
	 */
```

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +652,132 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+ *     2) NOT holding exclusive VM access
+ *     3) Dispatcher is idle (no task can be dispatched)
+ *     4) Can't externally enter shutdown from another source
+ */
```
missing description of params and return val.

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +652,132 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+ *     1) NOT holding exclusive VM access
+ *     2) Dispatcher is idle (no task can be dispatched)
+ *     3) Can't be in/enter shutdown
+ */
```
missing description of params and return val.

babsingh commented on gc/base/ParallelDispatcher.hpp:
```diff
@@ -116,10 +119,17 @@ class MM_ParallelDispatcher : public MM_BaseVirtual
 	
 	uintptr_t adjustThreadCount(uintptr_t maxThreadCount);
 	
+	virtual bool internalStartupThreads(uintptr_t workerThreadCount, uintptr_t maxWorkerThreadIndex);
```
missing function description

babsingh commented on gc/base/ParallelDispatcher.hpp:
```diff
@@ -116,10 +119,17 @@ class MM_ParallelDispatcher : public MM_BaseVirtual
 	virtual void shutDownThreads();
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	virtual bool expandThreadPool(MM_EnvironmentBase *env);
+	virtual void contractThreadPool(MM_EnvironmentBase *env, uintptr_t newThreadCount);
```
missing function description

dmitripivkine commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +652,132 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		omrthread_monitor_exit(_dispatcherMonitor);
+
+		/* Cleanup the dispatcher tables */
+		for(uintptr_t index = newThreadCount; index < _threadCountMaximum; index++) {
```
please add space before opening bracket

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +652,132 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+}
+
+/**
+ * Expand/fill the thread pool by starting up threads based on what H/W suppourts. This API is capped by the initial
+ * thread pool size, i.e we can't expand past _threadCountMaximumAtCheckpointStartup (currently don't support expanding dispatcher tables).
```
Exceeds 100 chars per line.

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +652,132 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+	uintptr_t preExpandThreadCount = _extensions->gcThreadCount;
+	uintptr_t newThreadCount = OMR_MIN(_threadCountMaximumAtCheckpointStartup, supportedGCThreadCount);
+
+	Trc_MM_ParallelDispatcher_expandThreadPool_params(newThreadCount, _threadCountMaximumAtCheckpointStartup, supportedGCThreadCount, preExpandThreadCount);
```
Too long (exceeds 100 chars). We should distribute it across multiple lines: https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md#function-parameter-list-does-not-fit-on-one-line.

babsingh commented on gc/base/standard/ConfigurationGenerational.cpp:
```diff
@@ -261,4 +261,47 @@ MM_ConfigurationGenerational::calculateDefaultRegionSize(MM_EnvironmentBase *env
+		extensions->concurrentScavengerBackgroundThreads = extensions->gcThreadCount;
+	}
+}
+#endif
```
```suggestion
#endif /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
```

babsingh commented on gc/base/standard/ConfigurationGenerational.cpp:
```diff
@@ -261,4 +261,47 @@ MM_ConfigurationGenerational::calculateDefaultRegionSize(MM_EnvironmentBase *env
+
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	initializeConcurrentScavengerThreadCount(env);
+#endif
```
```suggestion
#endif /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
```

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -294,26 +300,44 @@ MM_ParallelDispatcher::initialize(MM_EnvironmentBase *env)
+{
+	_threadShutdownCount = 0;
+
+	/* We may be starting the main thread at this point too */
```
avoid first person.
```suggestion
	/* The main thread may concurrently start at the same time. */
```

babsingh reviewed (changes requested):
Formatting feedback.

At some places, the function description is either missing, incomplete or incorrectly formatted.

Do not exceed 100 chars per line for comments and code.

For comments,
- there are missing endif comments;
- some don't end with a full-stop (inconsistent); and
- avoid first person and use third person.

@RSalman Please address all instances of the above feedback points only for the code that you have edited or added. I may have missed a few instances. Also, don't modify formatting for unchanged code.

RSalman commented:
Updated PR to remove the introduced assert in `shutdownThreads` and remove failure path in `startupthreads` (Failure path is handled by caller, startupThreads unnecessary calls shutdownThreads during failure), discussed with @amicic.

Formatting issue still to be resolved.

babsingh commented on gc/base/Configuration.cpp:
```diff
@@ -437,14 +437,22 @@ MM_Configuration::initializeGCThreadCount(MM_EnvironmentBase* env)
+	OMRPORT_ACCESS_FROM_OMRPORT(env->getPortLibrary());
+	uintptr_t threadCount = omrsysinfo_get_number_CPUs_by_type(OMRPORT_CPU_TARGET);
+
+	/* but not higher than maximum default */
```
this comments needs to be elaborated, and also the formatting needs to be fixed.

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +647,114 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+
+	Trc_MM_ParallelDispatcher_contractThreadPool_Entry(preShutdownThreadCount, newThreadCount);
+
+	/* Can't shutdown the main thread (Dispatcher didn't start it) */
```
```suggestion
	/* The main thread can't shutdown since the dispatcher didn't start it. */
```

babsingh commented on gc/base/ParallelDispatcher.cpp:
```diff
@@ -630,3 +647,114 @@ MM_ParallelDispatcher::reinitAfterFork(MM_EnvironmentBase *env, uintptr_t newThr
+		if (result) {
+			Assert_MM_true(_threadShutdownCount == (newThreadCount - 1));
+		} else {
+			/* We don't know how many threads started up prior to failing, infer from _threadShutdownCount */
```
use third person
```suggestion
			/* Infer from _threadShutdownCount to determine the number of threads that started up prior to failing.  */
```

babsingh commented on gc/base/ParallelDispatcher.hpp:
```diff
@@ -116,10 +119,53 @@ class MM_ParallelDispatcher : public MM_BaseVirtual
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+	/**
+	 * Expand/fill the thread pool by starting up threads based on what H/W supports.
+	 * This API is capped by the initial thread pool size, i.e we can't expand
+	 * past _threadCountMaximumAtCheckpointStartup (currently don't support expanding dispatcher tables).
```
use third person

babsingh commented on gc/base/standard/ConfigurationGenerational.cpp:
```diff
@@ -261,4 +261,47 @@ MM_ConfigurationGenerational::calculateDefaultRegionSize(MM_EnvironmentBase *env
+{
+	MM_GCExtensionsBase* extensions = env->getExtensions();
+
+	/* If not explicitly set, concurrent phase of CS runs with approx 1/4 the thread count (relative to STW phases thread count */
```
exceeds 100 chars

babsingh commented on gc/base/standard/Scavenger.hpp:
```diff
@@ -600,6 +600,13 @@ class MM_Scavenger : public MM_Collector
 	void calculateRecommendedWorkingThreads(MM_EnvironmentStandard *env);
 
+	/**
+	 * Sets the collector recommended thread count to UDATA_MAX (default value).
```
we should either consistently use full-stops or not. fix all instances.

babsingh commented:
jenkins build all

babsingh commented:
If the PR builds pass, I plan to merge this PR. @dmitripivkine Do you have any more feedback for this PR? If so, I won't merge.
